### PR TITLE
Add restless simulator tutorials

### PR DESCRIPTION
**Pull request by conradhaupt**
### Summary

This PR adds tutorials on how to use the restless simulator for simple qubit and qutrit circuits and with Qiskit Experiments.

### Details and comments

This PR adds tutorials which cover:
1. Running qubit circuits with restless circuit execution using the simulator.
2. Simulating qutrit circuits.
3. Standard circuit execution with the restless simulator.
4. Using the restless simulator with Qiskit Experiments.

It also includes a bug fixes for the simulator and package which would override certain options and made the package incompatible with Tox 4 and Qiskit Experiments 0.5.0. Tests for Python 3.7 were failing, and Qiskit has deprecated support. Because of this, I increased the minimum version to 3.8 in this PR.

### CODE REVIEW

File: docs/tutorials/1_introduction_to_simulator.ipynb [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "734365f3",
+   "metadata": {},
+   "source": [
+    "You can construct a restless simulator from the `QutritRestlessSimulator` class. To demonstrate how\n",
+    "the simulator works, we will create a series of single-qubit circuits with an initial $RX(\\pi/2)$\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/1_introduction_to_simulator.ipynb]: Addressed in a23b39e.

File: docs/tutorials/1_introduction_to_simulator.ipynb [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    "gate and an increasing number of identical $RX(\\pi/3)$ gates. We do not need to add measurements as\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/1_introduction_to_simulator.ipynb]: Addressed in a23b39e.

File: docs/tutorials/1_introduction_to_simulator.ipynb [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    "simulator = QutritRestlessSimulator()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "cbdf154d",
+   "metadata": {},
+   "source": [
+    "Here we create 10 circuits with increasing number of $RX(\\pi/3)$ gates.\n"
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/1_introduction_to_simulator.ipynb]: Addressed in a23b39e.

File: docs/tutorials/1_introduction_to_simulator.ipynb [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "2e2551fa",
+   "metadata": {},
+   "source": [
+    "As our circuits contain repeated applications of the $RX(\\pi/3)$ gate, we know how to compute the\n",
+    "ideal probabilities analytically. If we represent the probability to measure `\"0\"` $P_0[n]$ as a\n",
+    "function of the number of $RX(\\pi/3)$ gates $n$, we get the following cosine.\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/1_introduction_to_simulator.ipynb]: Addressed in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+   "metadata": {},
+   "source": [
+    "# Simulating Qutrit Circuits\n",
+    "\n",
+    "The `QutritRestlessSimulator` supports circuits with qutrit instructions. These instructions can be\n",
+    "unitary or general completely-positive trace-preserving maps. These two types are represented by the\n",
+    "`QutritUnitaryGate` and `QutritQuantumChannelOperation` classes respectively.\n",
+    "\n",
+    "This tutorial covers how to create qutrit circuits and how to simulate them. We use the same types\n",
+    "of circuits as in the introductory tutorial but with a leaky initial $RX(\\pi/2)$ gate."
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

File: docs/tutorials/3_standard_circuit_execution.ipynb [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Standard Circuit Execution with the Restless Simulator\n",
+    "\n",
+    "If we want, we can emulate qubit/qutrit reset with the `QutritRestlessSimulator` by modifying the\n",
+    "post-measurement transition matrix to implement reset. This allows us to execute qutrit circuits\n",
+    "with standard circuit execution.\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/3_standard_circuit_execution.ipynb]: Addressed in a23b39e.

File: docs/tutorials/3_standard_circuit_execution.ipynb [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+   "source": [
+    "## Create ideal and leaky circuits"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create leaky and ideal $RX(\\pi/2)$ gates"
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/3_standard_circuit_execution.ipynb]: Addressed in a23b39e.

File: docs/tutorials/3_standard_circuit_execution.ipynb [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+    "# Create simulator\n",
+    "simulator = QutritRestlessSimulator()"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create ideal and leaky circuits"
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/3_standard_circuit_execution.ipynb]: Addressed in a23b39e.

File: docs/tutorials/4_qiskit_experiments_with_simulator.ipynb [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/conrad/OneDrive/Documents/Work/IBM_Zurich_Internship_2022/Projects/restless-simulator/restless_simulator/simulator/qutrit_restless_simulator.py:717: UserWarning: Options field meas_level is not valid for this backend\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/4_qiskit_experiments_with_simulator.ipynb]: Yes. I've fixed this in a23b39e.

File: docs/tutorials/4_qiskit_experiments_with_simulator.ipynb [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "ExperimentEntryNotFound",
+     "evalue": "'Figure 0 not found.'",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/4_qiskit_experiments_with_simulator.ipynb]: Yes. I've fixed this in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+    "the simulator` tutorial, i.e., a leaky $RX(\\pi/2)$ gate followed by a sequence of ideal $RX(\\pi/3)$\n",
+    "gates. The leaky gate will reduce the contrast of our measurements as the $|2\\rangle$ population\n",
+    "will now be non-zero. The $RX(\\pi/3)$ gates will not interact with the $|2\\rangle$-state population\n",
+    "as they are ideal. The default measurement assignment for the simulator assigns the `\"1\"` label to\n",
+    "the $|2\\rangle$ state, which means that by default it cannot distinguish between $|1\\rangle$ and\n",
+    "$|2\\rangle$. This can be changed by setting the `meas_assignment_mat` option of the simulator.\n",
+    "\n",
+    "First we create a NumPy array of our leaky unitary gate as an ideal $RX(\\pi/2)$ rotation between\n",
+    "$|0\\rangle$ and $|1\\rangle$ and a small rotation from $|1\\rangle$ into the $|2\\rangle$ state. This\n",
+    "final rotation is controlled by the `leakage_rotation` variable.\n"
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As you can see, the measurements with a leaky initial $RX(\\pi/2)$ gate contaminate the measurement\n",
+    "results and give values that deviate from the ideal. In the next tutorial you will see how the\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+    "    0.5 * np.cos(np.pi / 3 * np.arange(N_CIRCUITS, step=0.1) + np.pi / 2) + 0.5\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As you can see, the measurements with a leaky initial $RX(\\pi/2)$ gate contaminate the measurement\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+   "source": [
+    "## Simulating Qutrit Circuits"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We simulate our circuits in exactly the same way as in the introductory tutorial. As we have not\n",
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

File: docs/tutorials/2_simulating_qutrit_circuits.ipynb [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that we have our leaky unitary, we create an instance of `QutritUnitaryGate` with our unitary\n",
+    "matrix. We then construct leaky circuits and use our leaky gate as the first gate. We also create\n",
+    "circuits with an ideal $RX(\\pi/2)$ gate against which we will compare our measurements.\n"
```
[the remark this answers is not in the document]
Comment (conradhaupt) [reply docs/tutorials/2_simulating_qutrit_circuits.ipynb]: Addressed in a23b39e.

**Comment by conradhaupt** [comment]
I've updated the notebooks based on the feedback and attempted to fix some bugs to make the tests pass. These fixes include:
- Updating tox.ini for Tox version 4.
- Fixing an incompatibility with Qiskit Experiments 0.5.0, relating to `Target`.
- Fixed some failing tests relating to dependencies, which required removing the matplotlib and Qiskit Aer requirements.

**Review by eggerdj — COMMENTED** [review]
3.7 removal looks ok

**Pull request merged**